File: DoohClick.Web/src/app/campaign/service/campaign.service.ts
import { Injectable } from '@angular/core';
import { WebApiService } from '../../shared/service/web-api.service';
import { MvGridParamOption } from '../../shared/model/param.model';
import {
  MvCampaign,
  MvCampaignFilterOptionParam,
  MvCampaignIdParam,
} from '../model/campaign.model';
import { map, Observable } from 'rxjs';
import { MvGridResponse, MvResponse } from '../../shared/model/response.model';

@Injectable({
  providedIn: 'root',
})
export class CampaignService {
  constructor(private api: WebApiService) {}

  getGrid(
    param: MvGridParamOption<MvCampaignFilterOptionParam>,
  ): Observable<MvResponse<MvGridResponse<MvCampaign>>> {
    return this.api.get('commercial/campaign/grid', param, true).pipe(
      map((response) => {
        if (response.data?.data) {
          response.data.data = response.data.data.map(this.mapDates);
        }
        return response;
      }),
    );
  }

  save(param: MvCampaign): Observable<MvResponse<MvCampaign>> {
    return this.api.post('commercial/campaign', param);
  }

  remove(param: MvCampaignIdParam): Observable<MvResponse<MvCampaignIdParam>> {
    return this.api.delete('commercial/campaign/{id}', param);
  }

  private mapDates(c: MvCampaign): MvCampaign {
    return {
      ...c,
      startDate: new Date(c.startDate),
      endDate: new Date(c.endDate),
      campaignFlight: c.campaignFlight?.map((f) => ({
        ...f,
        startDate: new Date(f.startDate),
        endDate: new Date(f.endDate),
      })),
    };
  }
}

File: DoohClick.API/DoohClick.Service/Application/commercial/campaign/CampaignService.cs

using DoohClick.DataAccess.Dapper;
using DoohClick.DataAccess.Data;
using DoohClick.DataAccess.Entity;
using DoohClick.Interface.Application.commercial.campaign;
using DoohClick.Interface.Shared.JsonSerializer;
using DoohClick.Model.Application.commercial.campaign;
using DoohClick.Model.Shared.Param;
using DoohClick.Model.Shared.Response;
using DoohClick.Service.Shared.Base;
using Microsoft.EntityFrameworkCore;
using Newtonsoft.Json;

namespace DoohClick.Service.Application.commercial.campaign
{
    public class CampaignService : BaseService, ICampaignService
    {
        private readonly AppDbContext _dbContext;
        public CampaignService(
            IDataAccessService dataAccessService,
            IJsonSerializer jsonSerializer,
            AppDbContext dbContext
            ) : base(dataAccessService, jsonSerializer)
        {
            _dbContext = dbContext;
        }

        public async Task<MvGridResponse<MvCampaign>?> GetGrid(MvGridParamOption<MvCampaignFilterOptionParam> param)
        {
            string result = await _dataAccessService.RetrievalProcedure("dbo.sp_campaign_sel", JsonConvert.SerializeObject(param));
            return _jsonSerializer.DeserializeObject<MvGridResponse<MvCampaign>>(result);
        }

        public async Task<List<MvCampaignScreenSchedule>?> GetSchedules(MvCampaignIdParam param)
        {
            string result = await _dataAccessService.RetrievalProcedure("dbo.sp_campaign_screen_schedule_sel", JsonConvert.SerializeObject(param));
            return _jsonSerializer.DeserializeObject<List<MvCampaignScreenSchedule>>(result);
        }

        public async Task<MvCampaign?> Save(MvCampaign param)
        {
            string result = await _dataAccessService.ActionProcedure("dbo.sp_campaign_tsk", JsonConvert.SerializeObject(param));
            return _jsonSerializer.DeserializeObject<MvCampaign>(result);
        }

        public async Task<MvCampaignScreenSchedule?> SaveSchedule(MvCampaignScreenScheduleParam param)
        {
            string result = await _dataAccessService.ActionProcedure("dbo.sp_campaign_screen_schedule_tsk", JsonConvert.SerializeObject(param));
            return _jsonSerializer.DeserializeObject<MvCampaignScreenSchedule>(result);
        }
        
        public async Task<MvCampaignIdParam?> Remove(MvCampaignIdParam param)
        {
            string result = await _dataAccessService.ActionProcedure("dbo.sp_campaign_del", JsonConvert.SerializeObject(param));
            return _jsonSerializer.DeserializeObject<MvCampaignIdParam>(result);
        }

        public async Task<MvCampaignIdParam?> Approve(MvCampaignIdParam param)
        {
            Campaign? result = await _dbContext.Campaigns
            .FirstOrDefaultAsync(c => c.Id == param.Id);

            if (result == null) return null;

            result.Status = "ACTIVE";
            result.ModifiedBy = param.UpdatedBy;
            result.ModifiedAt = DateTime.UtcNow;

            await _dbContext.SaveChangesAsync();

            return new MvCampaignIdParam { Id = result.Id };
        }
    }
}
